generals/remote/generalsio_client.py

```diff
@@ -0,0 +1,245 @@
+        self._status = "off"  # can be "off","game","lobby","queue"
+
+    @property
+    def queue_id(self):
+        if not self._queue_id:
+            raise GeneralsIOClientError("Queue ID is not set.\nIs agent in the game lobby?")
+
+        return self._queue_id
+
+    @property
+    def status(self):
+        return self._status
+
+    def _emit_receive(self, *args):
+        self.emit(*args)
+        return self.receive()
+
+    def register_agent(self, username: str) -> None:
+        """
+        Register Agent to GeneralsIO platform.
+        You can configure one Agent per `user_id`. `user_id` should be handled as secret.
+        :param user_id: secret ID of Agent
+        :param username: agent username, must be prefixed with `[Bot]`
+        """
+        event, response = self._emit_receive("set_username", (self.user_id, username))
+        if response:
+            # in case of success the response is empty
+            raise RegisterAgentError(response)
+
+    def join_private_lobby(self, queue_id: str) -> None:
+        """
+        Join (or create) private game lobby.
+        :param queue_id: Either URL or lobby ID number
+        """
+        self._status = "lobby"
+        self._emit_receive("join_private", (queue_id, self.user_id))
+        self._queue_id = queue_id
+
+    def join_game(self, force_start: bool = True) -> None:
+        """
+        Set force start if requested and wait for the game start.
+        :param force_start: If set to True, the Agent will set `Force Start` flag
+        """
+        self._status = "queue"
+        while True:
+            event, *data = self.receive()
+            self.emit("set_force_start", (self.queue_id, force_start))
+            if event == "game_start":
+                self._status = "game"
+                self._initialize_game(data)
+                self._play_game()
+                break
+
+    def _initialize_game(self, data: dict) -> None:
+        """
+        Triggered after server starts the game.
+        :param data: dictionary of information received in the beginning
+        """
+        self.game_state = GeneralsIOstate(data[0])
+
+    def _generate_action(self, observation: Observation) -> None:
+        """
+        Translate action from Agent to the server format.
+        :param action: dictionary representing the action
+        """
+        obs = observation.as_dict()
+        action = self.agent.act(obs)
+        if not action["pass"]:
+            source: np.ndarray = np.array(action["cell"])
+            direction = np.array(DIRECTIONS[action["direction"]].value)
+            split = action["split"]
+            destination = source + direction
+            # convert to index
+            source_index = source[0] * self.game_state.map[0] + source[1]
+            destination_index = destination[0] * self.game_state.map[0] + destination[1]
+            self.emit("attack", (int(source_index), int(destination_index), int(split)))
```

Comment:
that is just a presonal preferrence, but I would decouple sending the message from this method and move it to the `_play_game` method --> to have game message be handled in the single class.

Reply:
single class? you mean function?

Reply:
i think i see what you mean and it makes sense, will do that